File: src/lib/package-skill.ts
/**
 * `.skill` packaging: validate first, then collect files with the upstream
 * exclusion rules and write a ZIP whose structure is `<skill-name>/ + content`.
 *
 * Exclusion rules (ported as behaviour from the upstream package_skill.py):
 *   EXCLUDE_DIRS  = { "__pycache__", "node_modules" }  (any depth)
 *   EXCLUDE_GLOBS = { "*.pyc" }                        (any depth)
 *   EXCLUDE_FILES = { ".DS_Store" }                    (any depth)
 *   "evals/" is excluded only at the skill root.
 */

import { basename, dirname, join, relative, sep } from 'node:path'
import type { HostFs } from './fs-adapter.ts'
import { formatChecks, validateSkillDir, type ValidationResult } from './validate-skill.ts'
import { buildZip, type ZipEntry } from './zip.ts'

export const PACKAGE_EXCLUDE_DIRS: ReadonlySet<string> = new Set(['__pycache__', 'node_modules'])
export const PACKAGE_EXCLUDE_GLOBS: readonly string[] = ['*.pyc']
export const PACKAGE_EXCLUDE_FILES: ReadonlySet<string> = new Set(['.DS_Store'])
const SKILL_ROOT_EVALS = 'evals'

/** Simple glob matching supporting `*` and `?`. */
export function matchSimpleGlob(glob: string, name: string): boolean {
  const pattern = '^' + glob.replace(/\./g, '\\.').replace(/\*/g, '.*').replace(/\?/g, '.') + '$'
  return new RegExp(pattern).test(name)
}

/**
 * Apply the upstream exclusion rules to one path relative to the skill root.
 * `isDirectory` distinguishes directory-only rules (excluded dirs are pruned
 * during the walk, so their children never surface).
 */
export function shouldExcludeSkillPath(rel: string, isDirectory: boolean): boolean {
  const segments = rel.split('/').filter((segment) => segment.length > 0)
  if (segments.length === 0) return false
  const base = segments[segments.length - 1]!
  if (isDirectory) {
    if (PACKAGE_EXCLUDE_DIRS.has(base)) return true
    // evals/ is excluded only when it is a direct child of the skill root.
    if (segments.length === 1 && base === SKILL_ROOT_EVALS) return true
    return false
  }
  if (PACKAGE_EXCLUDE_FILES.has(base)) return true
  for (const glob of PACKAGE_EXCLUDE_GLOBS) {
    if (matchSimpleGlob(glob, base)) return true
  }
  return false
}

export interface CollectedFile {
  readonly abs: string
  readonly rel: string
}

export interface CollectedSkill {
  readonly name: string
  readonly files: readonly CollectedFile[]
}

/** Recursively collect packageable files under a skill directory. */
export async function collectSkillFiles(fs: HostFs, skillDir: string): Promise<CollectedSkill> {
  const name = basename(skillDir)
  const files: CollectedFile[] = []
  async function walk(current: string): Promise<void> {
    let entries: string[]
    try {
      entries = await fs.listDir(current)
    } catch {
      return
    }
    for (const entry of entries) {
      const abs = join(current, entry)
      const rel = relative(skillDir, abs).split(sep).join('/')
      let isDir = false
      let isFile = false
      try {
        const info = await fs.stat(abs)
        if (info === undefined) continue
        isDir = info.type === 'dir'
        isFile = info.type === 'file'
      } catch {
        // unreadable entry: skip
      }
      if (shouldExcludeSkillPath(rel, isDir)) continue
      if (isDir) {
        await walk(abs)
      } else if (isFile) {
        files.push({ abs, rel })
      }
    }
  }
  await walk(skillDir)
  return { name, files }
}

export interface SkillPackageResult {
  readonly ok: boolean
  readonly skillName: string
  readonly outPath: string
  readonly entries: readonly string[]
  readonly validation: ValidationResult
}

/**
 * Package a validated skill directory into `<skillName>.skill`. Throws when
 * validation fails (mirroring the upstream "refuse to package" behaviour).
 */
export async function packSkillDir(fs: HostFs, skillDir: string, outDir?: string): Promise<SkillPackageResult> {
  const validation = await validateSkillDir(fs, skillDir)
  if (!validation.ok) {
    throw new Error(`packSkill: validation failed:\n${formatChecks(validation.checks)}`)
  }
  const collected = await collectSkillFiles(fs, skillDir)
  if (collected.files.length === 0) {
    throw new Error('packSkill: no files to package')
  }

  const entries: ZipEntry[] = [{ name: `${collected.name}/`, data: new Uint8Array(0) }]
  for (const file of collected.files) {
    const content = await fs.readText(file.abs)
    entries.push({ name: `${collected.name}/${file.rel}`, data: new TextEncoder().encode(content) })
  }
  const zipBytes = buildZip(entries)

  const outPath = outDir !== undefined && outDir.length > 0
    ? join(outDir, `${collected.name}.skill`)
    : join(dirname(skillDir), `${collected.name}.skill`)
  await fs.mkdirp(dirname(outPath))
  await fs.writeBin(outPath, zipBytes)

  return {
    ok: true,
    skillName: collected.name,
    outPath,
    entries: entries.map((entry) => entry.name),
    validation,
  }
}

File: scripts/package_skill.py
#!/usr/bin/env python3
"""Package a DSH skill directory into a distributable `.skill` (zip) file.

Exclusion rules ported from the upstream claude-plugins-official skill-creator
`scripts/package_skill.py` (behaviour); the implementation is an independent
rewrite under the MIT license:

    EXCLUDE_DIRS  = {"__pycache__", "node_modules"}   (any depth)
    EXCLUDE_GLOBS = {"*.pyc"}                         (any depth)
    EXCLUDE_FILES = {".DS_Store"}                     (any depth)
    "evals/" is excluded only at the skill root.

Packaging first runs scripts/quick_validate.py's `validate_skill()`; invalid
skills refuse to package. The archive structure is `<skill-name>/ + contents`.
The produced file is written atomically (temp file + os.replace).
"""

from __future__ import annotations

import argparse
import fnmatch
import os
import sys
import tempfile
import zipfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from quick_validate import validate_skill  # noqa: E402

EXCLUDE_DIRS = {"__pycache__", "node_modules"}
EXCLUDE_GLOBS = {"*.pyc"}
EXCLUDE_FILES = {".DS_Store"}


def should_exclude(rel: str, is_dir: bool) -> bool:
    segments = [segment for segment in rel.split(os.sep) if segment]
    if not segments:
        return False
    base = segments[-1]
    if is_dir:
        if base in EXCLUDE_DIRS:
            return True
        # evals/ is excluded only when it is a direct child of the skill root.
        if len(segments) == 1 and base == "evals":
            return True
        return False
    if base in EXCLUDE_FILES:
        return True
    for glob in EXCLUDE_GLOBS:
        if fnmatch.fnmatch(base, glob):
            return True
    return False


def collect_entries(skill_root: str) -> list[tuple[str, str]]:
    """Return (rel_path, abs_path) pairs for every packageable file."""
    entries: list[tuple[str, str]] = []
    for dirpath, dirnames, filenames in os.walk(skill_root):
        rel_dir = os.path.relpath(dirpath, skill_root)
        kept: list[str] = []
        for name in dirnames:
            rel = name if rel_dir == "." else os.path.join(rel_dir, name)
            if not should_exclude(rel, True):
                kept.append(name)
        dirnames[:] = kept  # prune excluded directories from os.walk
        for name in filenames:
            rel = name if rel_dir == "." else os.path.join(rel_dir, name)
            if should_exclude(rel, False):
                continue
            entries.append((rel, os.path.join(dirpath, name)))
    entries.sort(key=lambda entry: entry[0])
    return entries


def package_skill(skill_dir: str, out_dir: str | None = None, out_path: str | None = None) -> str:
    ok, errors = validate_skill(skill_dir)
    if not ok:
        raise RuntimeError(
            "validation failed:\n" + "\n".join(f"- {error}" for error in errors)
        )
    root = Path(skill_dir)
    skill_name = root.name
    entries = collect_entries(skill_dir)
    if not entries:
        raise RuntimeError("no files to package")

    dest = out_path or str((Path(out_dir) if out_dir else root.parent) / f"{skill_name}.skill")
    dest_path = Path(dest)
    dest_path.parent.mkdir(parents=True, exist_ok=True)

    fd, tmp_name = tempfile.mkstemp(
        prefix=f".{skill_name}-", suffix=".tmp", dir=str(dest_path.parent)
    )
    os.close(fd)
    try:
        with zipfile.ZipFile(tmp_name, "w", compression=zipfile.ZIP_DEFLATED) as archive:
            archive.writestr(f"{skill_name}/", "")
            for rel, abs_path in entries:
                archive.write(abs_path, arcname=f"{skill_name}/{rel}")
        os.replace(tmp_name, dest)
    except BaseException:
        try:
            os.unlink(tmp_name)
        except OSError:
            pass
        raise
    return dest


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("path", help="skill directory to package")
    parser.add_argument("--out-dir", default=None, help="output directory for <name>.skill")
    parser.add_argument("--out", dest="out_path", default=None, help="exact output path")
    args = parser.parse_args(argv)

    try:
        dest = package_skill(args.path, out_dir=args.out_dir, out_path=args.out_path)
    except RuntimeError as exc:
        print(f"FAIL: {exc}", file=sys.stderr)
        return 1
    except OSError as exc:
        print(f"FAIL: {exc}", file=sys.stderr)
        return 1
    print(f"OK: {dest}")
    return 0


if __name__ == "__main__":
    sys.exit(main())